imagedephi/redact.py

```diff
@@ -1,74 +1,166 @@
-                for sub_ifds in tag_info.get("ifds", []):
-                    redact_tiff_tags(sub_ifds, tags_to_redact)
-
-
-def redact_one_image(tiff_info: TiffInfo, output_path: Path) -> None:
-    ifds = tiff_info["ifds"]
-    tags_to_redact = get_tags_to_redact()
-    redact_tiff_tags(ifds, tags_to_redact)
-    tifftools.write_tiff(tiff_info, output_path)
+class TiffMetadataRedactionPlan:
+    """
+    Represents a plan of action for redacting metadata from TIFF images.
+
+    The plan can be used for reporting to end users what steps will be made to redact their TIFF
+    images, and also executing the plan.
+    """
+
+    redaction_steps: dict[int, tuple[TiffMetadataRule, RuleSource]]
+    no_match_tags: list[tifftools.TiffTag]
+    image_data: TiffInfo
+    base_rules: list[TiffMetadataRule]
+    override_rules: list[TiffMetadataRule]
+
+    @staticmethod
+    def _iter_tiff_tag_entries(
+        ifds: list[IFD],
+    ) -> Generator[tuple[tifftools.TiffTag, IFD], None, None]:
+        for ifd in ifds:
+            for tag_id, entry in sorted(ifd["tags"].items()):
+                tag: tifftools.TiffTag = tifftools.constants.get_or_create_tag(
+                    tag_id, datatype=tifftools.Datatype[entry["datatype"]]
+                )
+                if not tag.isIFD():
+                    yield tag, ifd
+                else:
+                    # entry['ifds'] contains a list of lists
+                    # see tifftools.read_tiff
+                    for sub_ifds in entry.get("ifds", []):
+                        yield from TiffMetadataRedactionPlan._iter_tiff_tag_entries(sub_ifds)
+
+    def _add_tag_to_plan(self, tag: tifftools.TiffTag) -> None:
+        """Determine how to handle a given tag."""
+        for rule in self.override_rules:
+            if rule.is_match(tag):
+                self.redaction_steps[tag.value] = (rule, RuleSource.OVERRIDE)
+                return
+        for rule in self.base_rules:
+            if rule.is_match(tag):
+                self.redaction_steps[tag.value] = (rule, RuleSource.BASE)
+                return
+        self.no_match_tags.append(tag)
+
+    def _build_redaction_steps(self, ifds: list[IFD]) -> None:
+        for tag, _ in self._iter_tiff_tag_entries(ifds):
+            self._add_tag_to_plan(tag)
+
+    def __init__(
+        self, tiff_info: TiffInfo, base_rules: RuleSet, override_rules: RuleSet | None
+    ) -> None:
+        self.redaction_steps = {}
+        self.no_match_tags = []
+        self.image_data = tiff_info
+        self.base_rules = [
+            rule for rule in base_rules.rules[RuleFormat.TIFF] if isinstance(rule, TiffMetadataRule)
+        ]
+        if override_rules is None:
+            self.override_rules = []
+        else:
+            self.override_rules = [
+                rule
+                for rule in override_rules.rules[RuleFormat.TIFF]
+                if isinstance(rule, TiffMetadataRule)
+            ]
```

Comment:
Since there are no other rule types now, I think this filtering is premature, and it adds unnecessary code complexity. If we did have other rule types, lots of other things would probably fail anyway. I'd suggest just assigning it directly.

Reply:
I do like the idea of leaving out this sort of filtering for now. Perhaps `RuleSet.rules` could have a type of:
`dict[FileType, dict[RuleType, list[Rule]]`

The problem here is that `TiffMetadataRedactionPlan` expects its `base_rules` to be a list of `TiffMetadataRules`, so directly assigning makes mypy mad.

Reply:
See 19899e1 for changes here.

This filtering has been moved to the `RuleSet.get_tiff_metadata_rules` method. I *think* the filtering is necessary because the `RuleSet.rules` object contains lists of generic rules, so `mypy` doesn't like treating the list as a list of `TiffMetadataRules`.

I have changed that structure slightly so that all `TiffMetadataRule`s can live in their own list, instead of sharing with the future `TiffImageRule` list. 

Open to suggestions here.